src/storage/lock_manager/deadlock.rs

```diff
@@ -371,162 +506,101 @@ impl Inner {
+    fn handle_detect_locally(&self, tp: DetectType, txn_ts: u64, lock: Lock) {
+        let detect_table = &mut self.inner.borrow_mut().detect_table;
+        match tp {
+            DetectType::Detect => {
+                if let Some(deadlock_key_hash) = detect_table.detect(txn_ts, lock.ts, lock.hash) {
+                    self.waiter_mgr_scheduler
+                        .deadlock(txn_ts, lock, deadlock_key_hash);
+                }
             }
-            Ok(())
-        })
-        .map_err(|e| panic!("unexpected err: {:?}", e));
-        handle.spawn(timer);
-    }
-
-    fn initialize(&mut self, handle: &Handle) {
-        assert!(!self.is_initialized);
-        // Get leader info now because tokio_timer::Interval can't execute immediately.
-        let _ = Self::monitor_membership_change(&self.pd_client, &self.resolver, &self.inner);
-        self.schedule_membership_change_monitor(handle);
-        self.is_initialized = true;
+            DetectType::CleanUpWaitFor => {
+                detect_table.clean_up_wait_for(txn_ts, lock.ts, lock.hash)
+            }
+            DetectType::CleanUp => detect_table.clean_up(txn_ts),
+        }
     }
 
-    fn handle_detect(&self, handle: &Handle, tp: DetectType, txn_ts: u64, lock: Lock) {
-        let mut inner = self.inner.borrow_mut();
-        if inner.is_leader() {
-            match tp {
-                DetectType::Detect => {
-                    if let Some(deadlock_key_hash) = inner
-                        .detect_table
-                        .borrow_mut()
-                        .detect(txn_ts, lock.ts, lock.hash)
-                    {
-                        inner
-                            .waiter_mgr_scheduler
-                            .deadlock(txn_ts, lock, deadlock_key_hash);
-                    }
-                }
-                DetectType::CleanUpWaitFor => inner
-                    .detect_table
-                    .borrow_mut()
-                    .clean_up_wait_for(txn_ts, lock.ts, lock.hash),
-                DetectType::CleanUp => inner.detect_table.borrow_mut().clean_up(txn_ts),
-            }
+    /// Handles detect requests of itself.
+    fn handle_detect(&mut self, handle: &Handle, tp: DetectType, txn_ts: u64, lock: Lock) {
+        if self.is_leader() {
+            self.handle_detect_locally(tp, txn_ts, lock);
         } else {
             for _ in 0..2 {
-                if inner.leader_client.is_none() && inner.leader_info.is_some() {
-                    inner.reconnect_leader(handle);
+                // TODO: If the leader hasn't been elected, it requests Pd for
+                // each detect request. Maybe need flow control here.
+                //
+                // Refresh leader info when the connection to the leader is disconnected.
+                if self.leader_client.is_none() && !self.refresh_leader_info() {
+                    // Return if the leader doesn't exist.
+                    return;
```

Comment:
The detect request will be directly ignored here? May it cause any problem? Should here be a warn?

Reply:
It's a bug. I changed the `return` to `break` to log it.